### PR TITLE
Edge based userspace proxy

### PR DESCRIPTION
**Pull request by wojtek-t**
Second last PR from my changes to kube-proxy to make it event-based.

This is switching userspace proxy to be even-based similarly to what we already did with iptables.

### CODE REVIEW

**Comment by krmayankk** [comment]
@wojtek-t What does event based mean here ? Can you help understand or point to an issue which is being fixed here ? Is this also adding any new features to userspace based proxy ?

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -470,7 +470,7 @@ func NewProxyServer(config *componentconfig.KubeProxyConfiguration, cleanupAndEx
 			if err != nil {
 				return nil, fmt.Errorf("unable to create proxier: %v", err)
 			}
-			serviceHandler = proxierUserspace
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: can we get rid of serviceHandler now?

**Review by thockin — COMMENTED** [review]
I'm scared of changes here because this code is brittle, but this doesn't touch the bad parts, and I don't see anything wrong here.

**Comment by thockin** [comment]
@krmayankk this is just modernizing how API changes get passed in.

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (wojtek-t) [inline cmd/kube-proxy/app/server.go]: As soon as I will migrate windowsuserspace (will try to do it today).

**Comment by wojtek-t** [comment]
> @wojtek-t What does event based mean here ? Can you help understand or point to an issue which is being fixed here ? Is this also adding any new features to userspace based proxy ?

@krmayankk 
Event based here means that proxy is reacting on changes to particular services (add/modify/delete events) instead of grabing the whole snapshot (all services) and processing all of them from scratch all the time.
This doesn't introduce any new feature - it's just performance optimization.

**Comment by wojtek-t** [comment]
@thockin - thanks a lot!

@k8s-bot bazel test this

**Comment by wojtek-t** [comment]
@k8s-bot bazel test this

**Comment by wojtek-t** [comment]
I'm not able to make bazel run. Tried rebasing to see if this will help. Reapplying label.

**Pull request merged**